stdlib/public/core/StringProtocol.swift

```diff
@@ -17,9 +17,10 @@
 public protocol StringProtocol
   : BidirectionalCollection,
   TextOutputStream, TextOutputStreamable,
-  LosslessStringConvertible, ExpressibleByStringLiteral,
+  LosslessStringConvertible, ExpressibleByStringInterpolation,
   Hashable, Comparable
-  where Iterator.Element == Character, SubSequence : StringProtocol {
+  where Iterator.Element == Character, SubSequence : StringProtocol,
+        StringInterpolation == DefaultStringInterpolation {
```

Comment:
Out of curiosity, why is the new requirement needed? Do we want to make StringProtocol conform? Do we even want Substring to?

Reply:
It's not strictly necessary to support string interpolation in `Substring`s. But it's slightly weird for users that we don't, and it's really easy to do so now that we have `DefaultStringInterpolation`.

It would probably make it difficult to handle `StringProtocol`-conforming types which aren't built around `String`, though. If we're still thinking we might want to do that, we would probably need this to be something like `DefaultStringInterpolation<Base>`, where `Base` was an associated type that would be `String` for both `String` and `Substring`.